dart/dynamics/SharedLibraryIkFast.cpp

```diff
@@ -0,0 +1,265 @@
+  {
+    dterr << "Failed to load the symbol '<<" << symbolName
+          << "' from the file '" << fileName << "'.\n";
+    return false;
+  }
+
+  func = reinterpret_cast<FunctionType>(symbol);
+  // TODO(JS): Is there a way to check the compatibility of the symbol?
+
+  return true;
+}
+
+} // namespace (anonymous)
+
+//==============================================================================
+SharedLibraryIkFast::SharedLibraryIkFast(
+    InverseKinematics* ik,
+    const std::string& fileName,
+    const std::vector<std::size_t>& dofMap,
+    const std::vector<std::size_t>& freeDofMap,
+    const std::string& methodName,
+    const InverseKinematics::Analytical::Properties& properties)
+  : IkFast{ik, dofMap, freeDofMap, methodName, properties},
+    mFileName{fileName},
+    mSharedLibrary{nullptr},
+    mGetNumFreeParameters{nullptr},
+    mGetFreeParameters{nullptr},
+    mGetNumJoints{nullptr},
+    mGetIkRealSize{nullptr},
+    mGetIkType{nullptr},
+    mComputeIk{nullptr},
+    mGetKinematicsHash{nullptr},
+    mGetIkFastVersion{nullptr}
+{
+  auto lib = common::SharedLibrary::create(mFileName);
+
+  if (!lib)
+  {
+    dterr << "[SharedLibraryIkFast] Could not load dynamic library '"
+          << mFileName << "'. This SharedLibraryIkFast is invalid.\n";
+    return;
+  }
+
+  if (!loadFunction<IkFastFuncGetInt>(
+        lib, "GetNumFreeParameters", mFileName, mGetNumFreeParameters))
+      return;
+
+  if (!loadFunction<IkFastFuncGetIntPtr>(
+        lib, "GetFreeParameters", mFileName, mGetFreeParameters))
+    return;
+
+  if (!loadFunction<IkFastFuncGetInt>(
+        lib, "GetNumJoints", mFileName, mGetNumJoints))
+    return;
+
+  if (!loadFunction<IkFastFuncGetInt>(
+        lib, "GetIkRealSize", mFileName, mGetIkRealSize))
+    return;
+
+  if (!loadFunction<IkFastFuncGetInt>(
+        lib, "GetIkType", mFileName, mGetIkType))
+    return;
+
+  if (!loadFunction<IkFastFuncComputeIk>(
+        lib, "ComputeIk", mFileName, mComputeIk))
+    return;
+
+  if (!loadFunction<IkFastFuncGetConstCharPtr>(
+        lib, "GetKinematicsHash", mFileName, mGetKinematicsHash))
+    return;
+
+  if (!loadFunction<IkFastFuncGetConstCharPtr>(
+        lib, "GetIkFastVersion", mFileName, mGetIkFastVersion))
+    return;
+
+  mSharedLibrary = lib;
+}
+
+//==============================================================================
+auto SharedLibraryIkFast::clone(InverseKinematics* newIK) const
+    -> std::unique_ptr<GradientMethod>
+{
+  return dart::common::make_unique<SharedLibraryIkFast>(
+      newIK,
+      mFileName,
+      mDofs,
+      mFreeDofs,
+      getMethodName(),
+      getAnalyticalProperties());
+}
+
+//==============================================================================
+int SharedLibraryIkFast::getNumFreeParameters() const
+{
+  if (!mGetNumFreeParameters)
+  {
+    dterr << "[SharedLibraryIkFast::getNumFreeParameters] This SharedLibrary is "
+          << "invalid. Returning 0.\n";
+    return 0;
+  }
```

Comment:
Why are these checks necessary here, when you also check the return value of `loadFunction` in the constructor?

Reply:
This is because the constructor just tries to load the functions but does nothing even when it fails to load them. So it has to check whether it's actually loaded before being used.